src/mscorlib/src/System/Number.cs

```diff
@@ -1211,5 +1277,166 @@ internal unsafe struct NumberBuffer {
+            {
+                // Consume any leading or trailing whitespace
+                int start = 0;
+                int end = str.Length;
+                while (start < end && IsWhite(strPointer[start]))
+                {
+                    start++;
+                }
+                while (end > start && IsWhite(strPointer[end - 1]))
+                {
+                    end--;
+                }
+
+                if (start == end)
+                {
+                    // Nothing to parse
+                    return NumberTryParseResult.InvalidFormat;
+                }
+                if (strPointer[start] == '-')
+                {
+                    // Can't have a negative unsigned integer
+                    return NumberTryParseResult.Overflow;
+                }
+                for (int i = start; i < end; i++)
+                {
+                    int digitValue = 0;
+                    char c = strPointer[i];
+                    if (c >= '0' && c <= '9')
+                    {
+                        digitValue = c - '0';
+                    }
+                    else
+                    {
+                        // Non-digit char
+                        return NumberTryParseResult.InvalidFormat;
+                    }
+
+                    ulong previousResult = actualResult;
+                    actualResult = (actualResult * 10) + (ulong)digitValue;
+                    if (actualResult < previousResult)
+                    {
+                        // Overflow occured
+                        return NumberTryParseResult.Overflow;
+                    }
+                }
+                if (actualResult < minValue || actualResult > maxValue)
+                {
+                    return NumberTryParseResult.Overflow;
+                }
+                result = actualResult;
+                return NumberTryParseResult.Succeeded;
+            }
+        }
+
+        private unsafe static NumberTryParseResult TryParseSignedInvariantInteger(String str, long minValue, long maxValue, out long result)
+        {
+            result = 0;
+            if (str == null)
+            {
+                return NumberTryParseResult.NullString;
+            }
+
+            long actualResult = 0;
+            fixed (char* strPointer = str)
+            {
+                // Consume any leading or trailing whitespace
+                int start = 0;
+                int end = str.Length;
+                while (start < end && IsWhite(strPointer[start]))
+                {
+                    start++;
+                }
+                while (end > start && IsWhite(strPointer[end - 1]))
+                {
+                    end--;
+                }
+                
+                if (start == end)
+                {
+                    // Nothing to parse
+                    return NumberTryParseResult.InvalidFormat;
+                }
+
+                bool negative = strPointer[start] == '-';
```

Comment:
No you're right, good spot! Thanks.
I've added some more comprehensive tests for ParseNumber in dotnet/corefx#7407, but I agree the existing ones aren't brilliant